src/Calypso-SystemPlugins-Reflectivity-Queries-Tests/ClyActiveBreakpointsQueryTest.class.st

```diff
@@ -24,14 +24,16 @@ ClyActiveBreakpointsQueryTest >> installBreakpointInto: aMethod [
 ClyActiveBreakpointsQueryTest >> setUp [
 	super setUp.
 	
-	installedBreakpoints := OrderedCollection new
+	installedBreakpoints := OrderedCollection new.
+	ClyClassWithHalts compile: 'methodWithHalts1 self halt'.
```

Comment:
You perform code changes during tests in live environment. I think it is very bad practice and it should be avoided. If you want to "emulate methods" you need to create classes in your own test environment. This way it will not affect the live system state.

Reply:
That is done for a lot of thing that is meta. 
For instance testing metalinks implies recompiling.
To me, code changes already happen during live tests.
Can you specify why you think it is wrong?

> > I updated. The pragma is not compatible with Calypso tests
> 
> There is already support for this pragma. But tests were not adopted. Try following change:
> 
> ```smalltalk
> ClyActiveBreakpointsQueryTest>>createQuery
> 	^(ClyActiveBreakpointsQuery from: ClyClassScope of: ClyClassWithHalts in: environment)
> 		excludesHaltsForTesting: false.
> ```

Ah, thanks. I will use that if we give up on the dynamic method creation, but all the Calypso model is hard to grasp so I did not see that (nor took the time to find it I admit).

> > , so instead I generate halts and remove them after tests, so that there are no junk halts left in the code (which should be avoided in general...).
> 
> I do not see why it should be avoided. It is a good to have examples in image. And here is a live example of methods with halt, how they looks and behaves in the browser

Well, Calypso tests or framework are not the place to demonstrate methods with halts, it is rather in the documentation of the system.
And how they work in the browser, as users can just put a halt, they can see it live.

The problem is that rogue halts (without the pragma) are going to halt if executed on the base system (which seems wrong?) and additionaly, it is interfering with debugging tools that we are building (and I don't see why it should).

What I mean is that halts without a specification that they are for tests or example should be avoided in the base image. Maybe we could use a pragma <example> that tools can filter?

Reply:
> That is done for a lot of thing that is meta.
> For instance testing metalinks implies recompiling.
> To me, code changes already happen during live tests.
> Can you specify why you think it is wrong?

Running those tests will mark code to be dirty. After image restart it will show lost changes. 
It is about side effect which tests produce. It is good practice to avoid it

Reply:
> Well, Calypso tests or framework are not the place to demonstrate methods with halts, it is rather in the documentation of the system.

I agree. It would be good to have separate reusable package with such examples.

> And how they work in the browser, as users can just put a halt, they can see it live.

Examples demonstrate what is possible in the system. For "just put a halt" you need to know about it. Especially when you are from another language you have no idea about "self halt".
Documentation is a good place of course but it is not a live thing but the real code is.

Reply:
> ```smalltalk
> ```smalltalk
> ClyActiveBreakpointsQueryTest>>createQuery
> 	^(ClyActiveBreakpointsQuery from: ClyClassScope of: ClyClassWithHalts in: environment)
> 		excludesHaltsForTesting: false.
> ```
> 
> 
> 
> Ah, thanks. I will use that if we give up on the dynamic method creation, but all the Calypso model is hard to grasp so I did not see that (nor took the time to find it I admit).

Sure it can be difficult to find features. But I'm curious here it is about a class under broken test. Did you try debug test before making conclusion?

Reply:
> > Well, Calypso tests or framework are not the place to demonstrate methods with halts, it is rather in the documentation of the system.
> 
> I agree. It would be good to have separate reusable package with such examples.
> 
> > And how they work in the browser, as users can just put a halt, they can see it live.
> 
> Examples demonstrate what is possible in the system. For "just put a halt" you need to know about it. Especially when you are from another language you have no idea about "self halt".
> Documentation is a good place of course but it is not a live thing but the real code is.

True, that is why those users will look at the documentation or put breakpoints as in mainstream languages by clicking on the code :p

Reply:
> > That is done for a lot of thing that is meta.
> > For instance testing metalinks implies recompiling.
> > To me, code changes already happen during live tests.
> > Can you specify why you think it is wrong?
> 
> Running those tests will mark code to be dirty. After image restart it will show lost changes.
> It is about side effect which tests produce. It is good practice to avoid it

Hmm true. But that is the only way to test some things (again, metalinks).
I will try your solution with the pragma and the test configuration, but its impossible to do anything for all other code that does the same thing.